From c9f8cecbcb0dfbf43332eaedcd2bfb5fc482285c Mon Sep 17 00:00:00 2001
From: Tomas Mraz <tomas@openssl.foundation>
Date: Tue, 6 Oct 2026 18:44:19 +0200
Subject: [PATCH] rsa_multiprime_keygen(): Fix leak of tmp if stack insertion
 fails

Reviewed-by: Daniel Kubec <kubec@openssl.foundation>
Reviewed-by: Richard Levitte <levitte@openssl.org>
Reviewed-by: Milan Broz <mbroz@openssl.org>
Reviewed-by: Bob Beck <beck@openssl.org>
Merge-date: Wed Oct  7 07:39:03 2026
Merged-from: https://github.com/openssl/openssl/pull/33129
(cherry picked from commit 35b8874b1d45aed1a5a604cca75d7af3c1c72067)
---
 crypto/rsa/rsa_gen.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/crypto/rsa/rsa_gen.c b/crypto/rsa/rsa_gen.c
index 777a4854..12eecf7b 100644
--- a/crypto/rsa/rsa_gen.c
+++ b/crypto/rsa/rsa_gen.c
@@ -450,8 +450,10 @@ static int rsa_multiprime_keygen(RSA *rsa, int bits, int primes,
             tmp = BN_dup(prime);
             if (tmp == NULL)
                 goto err;
-            if (!sk_BIGNUM_insert(factors, tmp, sk_BIGNUM_num(factors)))
+            if (!sk_BIGNUM_insert(factors, tmp, sk_BIGNUM_num(factors))) {
+                BN_clear_free(tmp);
                 goto err;
+            }
             continue;
         }
 
@@ -520,8 +522,10 @@ static int rsa_multiprime_keygen(RSA *rsa, int bits, int primes,
         tmp = BN_dup(prime);
         if (tmp == NULL)
             goto err;
-        if (!sk_BIGNUM_insert(factors, tmp, sk_BIGNUM_num(factors)))
+        if (!sk_BIGNUM_insert(factors, tmp, sk_BIGNUM_num(factors))) {
+            BN_clear_free(tmp);
             goto err;
+        }
     }
 
     if (BN_cmp(rsa->p, rsa->q) < 0) {
